File: backend/payment/src/main/java/com/bbmovie/payment/controller/PaymentController.java
package com.bbmovie.payment.controller;

import com.bbmovie.payment.dto.*;
import com.bbmovie.payment.dto.request.PaymentRequest;
import com.bbmovie.payment.dto.request.CallbackRequestContext;
import com.bbmovie.payment.dto.request.RefundRequest;
import com.bbmovie.payment.dto.response.PaymentCreationResponse;
import com.bbmovie.payment.dto.response.PaymentVerificationResponse;
import com.bbmovie.payment.dto.response.RefundResponse;
import com.bbmovie.payment.service.PaymentService;
import com.bbmovie.payment.service.I18nService;
import jakarta.servlet.http.HttpServletRequest;
import lombok.extern.log4j.Log4j2;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

import static com.bbmovie.payment.utils.PaymentProviderPayloadUtil.*;

@Log4j2
@RestController
@RequestMapping("/api/payment")
public class PaymentController {

    private final PaymentService paymentService;
    private final I18nService i18nService;

    @Autowired
    public PaymentController(PaymentService paymentService, I18nService i18nService) {
        this.paymentService = paymentService;
        this.i18nService = i18nService;
    }

    @PostMapping("/initiate")
    public ResponseEntity<ApiResponse<PaymentCreationResponse>> initiatePayment(
            @SuppressWarnings("unused") @RequestHeader("Authorization") String jwtToken, // This Will be used to pass the user id
            @RequestBody PaymentRequest request, HttpServletRequest httpServletRequest
    ) {
        try {
            if (request.getExpiresInMinutes() == null) {
                request.setExpiresInMinutes(30); // default expiry window
            }
            PaymentCreationResponse response = paymentService.createPayment(request.getProvider(), request, httpServletRequest);
            return ResponseEntity.ok(ApiResponse.success(response));
        } catch (Exception e) {
            log.error("Error processing payment: {}", e.getMessage());
            return ResponseEntity.badRequest().body(ApiResponse.error(i18nService.getMessage("payment.error.processing")));
        }
    }

    /**
     *  <b>WARN</b>
     *  <p> This endpoint uses for testing </p>
     *  <p> In prod should redirect to FE instead of this endpoint, then FE makes post-request</p>
     */
    @GetMapping("/{provider}/callback")
    public ResponseEntity<ApiResponse<PaymentVerificationResponse>> handleProviderCallbackGet(
            @RequestParam Map<String, String> params, HttpServletRequest request, @PathVariable String provider
    ) {
        log.info("\nReceived payment callback from provider: {}.\n {}", provider,params);
        PaymentVerificationResponse verification = paymentService.handleCallback(provider, params, request);
        if (verification.isValid()) {
            return ResponseEntity.ok(ApiResponse.success(verification));
        }
        return ResponseEntity.status(HttpStatus.NOT_ACCEPTABLE).body(ApiResponse.fail(verification));
    }

    // For Zalopay, the server must return a minimal JSON with return_code/return_message per provider docs.
    @PostMapping("/zalopay/callback")
    public ResponseEntity<Map<String, String>> handleProviderCallbackPost(
            @RequestBody Map<String, String> body, HttpServletRequest request
    ) {
        log.info("\nReceived payment callback from ZALOPAY.\n {}", body);
        PaymentVerificationResponse verification = paymentService.handleCallback("zalopay", body, request);
        @SuppressWarnings("unchecked") // The data has been made sure to match
        Map<String, String> providerPayload = stringToJson(verification.getProviderResponse().toString(), Map.class);
        if (verification.isValid()) {
            return ResponseEntity.ok(providerPayload);
        }
        return ResponseEntity.status(HttpStatus.NOT_ACCEPTABLE).body(providerPayload);
    }

    @PostMapping("/{provider}/ipn")
    public ResponseEntity<ApiResponse<PaymentVerificationResponse>> handleGenericIpn(
            @PathVariable String provider, @RequestParam Map<String, String> params,
            HttpServletRequest request
    ) {
        CallbackRequestContext ctx = createContext(request, params, HttpMethod.POST);
        try {
            PaymentVerificationResponse verification = paymentService.handleIpn(provider, ctx);
            if (verification.isValid()) {
                return ResponseEntity.ok(ApiResponse.success(verification));
            }
            return ResponseEntity.status(HttpStatus.NOT_ACCEPTABLE).body(ApiResponse.fail(verification));
        } catch (UnsupportedOperationException ex) {
            return ResponseEntity.status(HttpStatus.NOT_IMPLEMENTED).body(ApiResponse.error(i18nService.getMessage("payment.error.ipn.unsupported")));
        }
    }

    @PostMapping("/{provider}/webhook")
    public ResponseEntity<ApiResponse<PaymentVerificationResponse>> handleGenericWebhook(
            @PathVariable String provider, @RequestBody String payload,
            @RequestHeader Map<String, String> headers, HttpServletRequest request
    ) {
        CallbackRequestContext ctx = createContext(request, payload, HttpMethod.POST, headers);
        try {
            PaymentVerificationResponse verification = paymentService.handleWebhook(provider, ctx);
            if (verification.isValid()) {
                return ResponseEntity.ok(ApiResponse.success(verification));
            }
            return ResponseEntity.status(HttpStatus.NOT_ACCEPTABLE).body(ApiResponse.fail(verification));
        } catch (UnsupportedOperationException ex) {
            return ResponseEntity.status(HttpStatus.NOT_IMPLEMENTED).body(ApiResponse.error(i18nService.getMessage("payment.error.webhook.unsupported")));
        }
    }

    @PostMapping("/refund")
    public ResponseEntity<ApiResponse<RefundResponse>> refundPayment(
        @SuppressWarnings("unused") @RequestHeader("Authorization") String jwtToken, // This Will be used to pass the user id
        @RequestBody RefundRequest request, 
        HttpServletRequest httpServletRequest
    ) {
        RefundResponse response = paymentService.refundPayment(String.valueOf(request.getPaymentId()), httpServletRequest);
        return ResponseEntity.ok(ApiResponse.success(response));
    }

    @GetMapping("/query")
    public ResponseEntity<ApiResponse<Object>> queryPaymentFromProvider(
        @SuppressWarnings("unused") @RequestHeader("Authorization") String jwtToken, // This Will be used to pass the user id
        @RequestParam String id
    ) {
        return ResponseEntity.ok(ApiResponse.success(paymentService.queryPayment(id)));
    }
}

File: backend/payment/src/main/kotlin/com/bbmovie/payment/controller/PaymentController.kt
package com.bbmovie.payment.controller

import com.bbmovie.payment.entity.PaymentStatus
import com.bbmovie.payment.service.payment.PaymentService
import com.bbmovie.payment.service.payment.dto.ApiResponse
import com.bbmovie.payment.service.payment.dto.PaymentRequest
import com.bbmovie.payment.service.payment.dto.PaymentRequestDto
import com.bbmovie.payment.service.payment.dto.PaymentResponse
import com.bbmovie.payment.service.payment.dto.PaymentVerification
import com.bbmovie.payment.service.payment.dto.RefundRequestDto
import com.bbmovie.payment.service.payment.dto.RefundResponse
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/payment")
class PaymentController
@Autowired constructor(private val paymentService: PaymentService) {

    private val log = LoggerFactory.getLogger(this.javaClass)

    @PostMapping("/initiate")
    fun initiatePayment(
        @RequestBody request: PaymentRequestDto,
        httpServletRequest: HttpServletRequest
    ): ResponseEntity<ApiResponse<PaymentResponse>> {
        try {
            val paymentRequest = PaymentRequest()
            paymentRequest.amount = request.amount
            paymentRequest.currency = request.currency
            paymentRequest.userId = request.userId
            paymentRequest.orderId = request.orderId
            val response: PaymentResponse? = paymentService.processPayment(
                request.provider as String, paymentRequest, httpServletRequest
            )
            return ResponseEntity.ok(ApiResponse.success(response))
        } catch (e: Exception) {
            log.error("Error processing payment: {}", e.message)
            return ResponseEntity.badRequest().body(
                ApiResponse.error(PaymentResponse(null, PaymentStatus.FAILED, e.message).toString())
            )
        }
    }

    @GetMapping("/vnpay/callback")
    fun handleVNPayCallback(
        @RequestParam params: MutableMap<String, String>,
        httpServletRequest: HttpServletRequest
    ): ResponseEntity<ApiResponse<PaymentVerification>> {
        try {
            val verification: PaymentVerification = paymentService.verifyPayment(
                "vnpayProvider", params, httpServletRequest
            )
            return ResponseEntity.ok(ApiResponse.success(verification))
        } catch (e: Exception) {
            log.error("Error processing VNPay callback: {}", e.message)
            return ResponseEntity.badRequest().body(
                ApiResponse.error(PaymentVerification(false, null).toString())
            )
        }
    }

    @PostMapping("/webhook")
    fun handleWebhook(
        @RequestHeader("X-Payment-Provider") provider: String,
        @RequestBody paymentData: MutableMap<String, String>,
        httpServletRequest: HttpServletRequest
    ): ResponseEntity<ApiResponse<PaymentVerification>> {
        try {
            val verification: PaymentVerification = paymentService.verifyPayment(provider, paymentData, httpServletRequest)
            return ResponseEntity.ok(ApiResponse.success(verification))
        } catch (e: Exception) {
            log.error("Error processing webhook: {}", e.message)
            return ResponseEntity.badRequest().body(
                ApiResponse.error(PaymentVerification(false, null).toString())
            )
        }
    }

    @PostMapping("/refund")
    fun refundPayment(
        @RequestBody request: RefundRequestDto,
        httpServletRequest: HttpServletRequest
    ): ResponseEntity<ApiResponse<RefundResponse>> {
        try {
            val response: RefundResponse? = paymentService.refundPayment(
                request.provider as String, request.paymentId as String, httpServletRequest
            )
            return ResponseEntity.ok(ApiResponse.success(response))
        } catch (e: Exception) {
            log.error("Error processing refund: {}", e.message)
            return ResponseEntity.badRequest().body(
                ApiResponse.error(RefundResponse(null, PaymentStatus.FAILED.getStatus).toString())
            )
        }
    }
}